File: packages/stride_health/android/src/main/kotlin/com/projectstride/stride_health/StrideHealthPlugin.kt
package com.projectstride.stride_health

import android.app.Activity
import android.content.Intent
import androidx.health.connect.client.PermissionController
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.PluginRegistry
import kotlin.coroutines.resume
import kotlinx.coroutines.suspendCancellableCoroutine

/**
 * Project Stride — Android step integration.
 *
 * Registration, the Pigeon boundary, and the foreground activity the permission
 * sheet needs. Nothing else, and deliberately nothing else: there is no
 * background registration, no worker, no service, and no passive-monitoring
 * subscription anywhere in this plugin. S-01A is foreground synchronization;
 * S-01B is blocked on a real persistence coordinator (DECISIONS/0013,
 * DECISIONS/0014).
 *
 * See DECISIONS/0010_CROSS_PLATFORM_STACK.md.
 */
class StrideHealthPlugin : FlutterPlugin, ActivityAware, PluginRegistry.ActivityResultListener {

    private var adapter: HealthConnectAdapter? = null
    private var activity: Activity? = null
    private var activityBinding: ActivityPluginBinding? = null

    /** The in-flight permission request, if the sheet is open. */
    private var pendingPermission: ((Set<String>?) -> Unit)? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        val created = HealthConnectAdapter(
            HealthConnectStepSource(binding.applicationContext, requester)
        )
        adapter = created
        HealthHostApi.setUp(binding.binaryMessenger, created)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        HealthHostApi.setUp(binding.binaryMessenger, null)
        // The keying salt is the app's device-bound identity, held here only
        // for the lifetime of the attachment. Dropping it on detach is what
        // makes "in memory only" true rather than merely intended -- and it
        // means the next attachment must install it again, which is the
        // fail-closed path.
        adapter?.dispose()
        adapter = null
    }

    // -----------------------------------------------------------------------
    // ActivityAware -- the permission sheet, and nothing else
    // -----------------------------------------------------------------------

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activityBinding = binding
        activity = binding.activity
        binding.addActivityResultListener(this)
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) =
        onAttachedToActivity(binding)

    override fun onDetachedFromActivityForConfigChanges() = onDetachedFromActivity()

    override fun onDetachedFromActivity() {
        activityBinding?.removeActivityResultListener(this)
        activityBinding = null
        activity = null
        // A sheet whose activity went away has no answer. Resolved as "could
        // not ask" rather than as a denial: the game never records a question
        // it failed to put as a no.
        pendingPermission?.invoke(null)
        pendingPermission = null
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
        if (requestCode != PERMISSION_REQUEST_CODE) return false
        val resume = pendingPermission ?: return false
        pendingPermission = null
        val granted = try {
            PermissionController.createRequestPermissionResultContract()
                .parseResult(resultCode, data)
        } catch (_: Throwable) {
            null
        }
        resume(granted)
        return true
    }

    private val requester = object : PermissionRequester {
        override suspend fun request(permissions: Set<String>): Set<String>? {
            val host = activity ?: return null
            if (pendingPermission != null) return null
            return suspendCancellableCoroutine { continuation ->
                pendingPermission = { granted -> continuation.resume(granted) }
                try {
                    val intent = PermissionController
                        .createRequestPermissionResultContract()
                        .createIntent(host, permissions)
                    host.startActivityForResult(intent, PERMISSION_REQUEST_CODE)
                } catch (_: Throwable) {
                    pendingPermission = null
                    continuation.resume(null)
                }
                continuation.invokeOnCancellation { pendingPermission = null }
            }
        }
    }

    private companion object {
        /** Arbitrary, and namespaced only by being this plugin's own. */
        const val PERMISSION_REQUEST_CODE = 0x53_48_01
    }
}

File: packages/stride_health/ios/stride_health/Sources/stride_health/StrideHealthPlugin.swift
import Flutter
import UIKit

/// Project Stride — iOS step integration.
///
/// S-01A scope: registration, the Pigeon boundary, and the FOREGROUND
/// HealthKit reader in `HealthKitStepStore`.
///
/// There is deliberately no `HKObserverQuery`, no `enableBackgroundDelivery`,
/// and no background mode registered here. Background synchronization is S-01B
/// and is blocked on a real persistence coordinator — see `DECISIONS/0014`.
///
/// See DECISIONS/0010_CROSS_PLATFORM_STACK.md.
public class StrideHealthPlugin: NSObject, FlutterPlugin {

    private static var adapter: HealthKitAdapter?

    public static func register(with registrar: FlutterPluginRegistrar) {
        // A fresh adapter per registration, and therefore no keying salt until
        // the app installs one. That is the fail-closed path: `fetchSteps`
        // refuses with `originKeyingUnconfigured` until `installOriginKeying`
        // has succeeded, because observations keyed under no salt would re-key
        // every origin and grant the retention window a second time.
        adapter?.forgetOriginKeying()
        let created = HealthKitAdapter()
        adapter = created
        HealthHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: created)

        // Published so the engine has somewhere to deliver the detach
        // callback. Without a published instance `detachFromEngineForRegistrar:`
        // is never sent, and the salt would outlive the attachment it belongs
        // to — which is the difference between "in memory for the lifetime of
        // the attachment" as a claim and as a fact.
        registrar.publish(StrideHealthPlugin())
    }

    /// Drops the keying salt when the engine goes away.
    ///
    /// After this, `fetchSteps` is fail-closed again with
    /// `originKeyingUnconfigured` until the app installs the identity afresh.
    /// Nothing is written anywhere on the way out: there is no native store to
    /// flush, by design.
    public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
        StrideHealthPlugin.adapter?.forgetOriginKeying()
        StrideHealthPlugin.adapter = nil
        HealthHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: nil)
    }
}
